### ISSUE TITLE
Support view() on batch dimensions for non-contiguous tensors?

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#3653]
```python
a = torch.Tensor(16, 8, 32, 64)
a.view(-1, 32, 64)                   # works
a.transpose(-1, -2).view(-1, 64, 32) # doesn't
a.view(-1, 32, 64).transpose(-1, -2) # works but doesn't fit some interfaces, i.e. sometimes we want to do view as the last operation
```

Such view's are needed to implement 4D+ `bmm` that can treat all dimensions except last two as batch dimensions (similarly to Linear module's behavior). Unless I move transpose inside the `bmm` func (which would not match the existing interface but well), an extra `contiguous` call is needed.

Does it make sense to support such `view` call? On one hand it breaks invariant that `view` always returns a contiguous tensor, on the other side the situation for several batch dimensions may be common.

Earlier discussed in https://github.com/pytorch/pytorch/issues/764

**Comment by apaszke** [comment]
There's a good reason for the `view` invariant - most of such reshapes are impossible to pull off using stride tricks if then tensor isn't contiguous. On the other hand, making it contiguous inside `view` would mean that sometimes the returned tensor shares storage with input, and sometimes doesn't. This is important for cases like these: `x.view(-1)[::x.size(1) + 1] += c` (add `c` to diagonal of matrix `x`). If you know/suspect that tensors might sometimes be non-contiguous just add `.contiguous()` before `.view()` it's a no-op if the tensor already is contiguous

**Comment by fmassa** [comment]
I think Vadim's point is that if the subspace of the non-contiguous tensor that we are modifying is contiguous, then the stride tricks should work, and we could reduce some of the restrictions in `view`.
This is similar in spirit with the `unsqueeze` function, that required dedicated implementation because of the restrictions in `view`.
Not sure if there are edge cases there though, and might just be safer/better to have everything uniform.

**Comment by apaszke** [comment]
Ohhh I haven't noticed this, my bad. What "works but doesn't fit some interfaces" means? I think it should give results equivalent to the previous line (if it worked). I think relaxing the constraint on `view` to be able to work with certain classes of non-contiguous tensors would be reasonable

**Comment by vadimkantorov** [comment]
Yep, you got me right :)

By "works but doesn't fit some interfaces" I meant that `bmm` interface accepts `BxPxQ` and `BxQxR`, so if I'm implementing a generalized version that flattens some first dimensions, then the inputs must already be properly transposed by caller. And then there's this problem that we can't flatten first dimensions of non-contiguous tensors.

**Comment by apaszke** [comment]
I'm not sure if I understand, can't you just call `contiguous()` inside your function (this doesn't change the interface), and only then flatten the leading dimensions? This will be a no-op if user gave you a properly transposed tensor, and will fix the problem otherwise

**Comment by vadimkantorov** [comment]
That's what I currently do.

Though (in my code), the tensor will always be non-contiguous and an unnecessary copy will happen, since transpose will have to be applied first:

```python
import torch
def bmm(A, B):
    print(A.is_contiguous(), B.is_contiguous())
    return torch.bmm(A.contiguous().view(-1, A.size(-2), A.size(-1)), B.contiguous().view(-1, B.size(-2), B.size(-1))).view(A.size()[:-1] + (-1, ))

a = torch.Tensor(16, 8, 32, 64)
bmm(a, a.transpose(-1, -2)) # prints (True, False)
```

**Comment by SsnL** [comment]
working on this

**Comment by soumith** [comment]
closed via #4062
